### PR TITLE
Extract metric to own module

### PR DESCRIPTION
**Pull request by bivas**
Remove all metrics reporting from main module.
At the moment, we implement opencencus metrics reporting using the `CheckListener` solution. 

*New API*
- Added `HealthListener` to enable events when results are updated
- It's now possible to register multiple `CheckListener`s
- `Option` API now supports registering `CheckListener`s and `HealthListener`s
- OpenCencus Specific
   - Ability to set metrics classification when creating listener

**Breaking Changes**
- Added `OnCheckRegistered` to `CheckListener` - this will enable actions when check is registered with the service. e.g. resetting metrics in opencencus

Discussion in #24

### CODE REVIEW

File: utils.go [hunk 1]
```diff
@@ -9,3 +9,11 @@ func allHealthy(results map[string]Result) (healthy bool) {
 
 	return true
 }
+
+func copyResultMap(result map[string]Result) map[string]Result {
+	newMap := make(map[string]Result)
```
Comment (eranharel) [inline utils.go:14]: I'd add the size `len(results)`

File: utils.go [hunk 2]
```diff
@@ -9,3 +9,11 @@ func allHealthy(results map[string]Result) (healthy bool) {
 
 	return true
 }
+
+func copyResultMap(result map[string]Result) map[string]Result {
```
Comment (eranharel) [inline utils.go:13]: result**s**

Comment (eranharel) [inline utils.go:13]: There's an health.Results() func. Why do we need this util too?

File: health.go [hunk 3]
```diff
@@ -141,8 +147,8 @@ func (h *health) DeregisterAll() {
 	h.lock.RLock()
 	defer h.lock.RUnlock()
 
-	for k := range h.checkTasks {
-		h.Deregister(k)
+	for _, task := range h.checkTasks {
+		task.stopChan <- true
```
Comment (eranharel) [inline health.go:151]: You didn't like the code reuse?

File: opencencus-check-listener/options.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+// WithReadinessClassification sets the classification to "readiness"
+func WithReadinessClassification() Option {
+	return func(listener *CheckListener) {
+		listener.classification = "readiness"
+	}
+}
+
+// WithSetupClassification sets the classification to "setup"
+func WithSetupClassification() Option {
```
Comment (eranharel) [inline opencencus-check-listener/options.go:26]: Did you mean to call it *startup*?

File: opencencus-check-listener/types.go [hunk 5]
```diff
@@ -1,41 +1,48 @@
...
+	keyClassification, _ = tag.NewKey("classification")
 
 	mCheckStatus   = stats.Int64("health/status", "An health status (0/1 for fail/pass)", "pass/fail")
 	mCheckDuration = stats.Float64("health/execute_time", "The time it took to execute a checks in ms", "ms")
 
 	// ViewCheckExecutionTime is the checks execution time aggregation tagged by check name
 	ViewCheckExecutionTime = &view.View{
 		Measure:     mCheckDuration,
-		TagKeys:     []tag.Key{keyCheck},
+		TagKeys:     []tag.Key{keyCheck, keyClassification},
```
Comment (eranharel) [inline opencencus-check-listener/types.go:29]: do we get extra rows when classification is empty?

File: opencencus-check-listener/utils.go [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package opencencus
+
+import gosundheit "github.com/AppsFlyer/go-sundheit"
+
+func allHealthy(results map[string]gosundheit.Result) (healthy bool) {
```
Comment (eranharel) [inline opencencus-check-listener/utils.go:5]: I guess we need a cleanup iteration and refactor `map[string]gosundheit.Result` into a type, and add this fucntion there

File: options.go [hunk 7]
```diff
@@ -3,30 +3,20 @@ package gosundheit
 type Option func(*health)
 
 // WithCheckListener allows you to listen to check start/end events
-func WithCheckListener(listener CheckListener) Option {
+func WithCheckListener(listener ...CheckListener) Option {
```
Comment (eranharel) [inline options.go:6]: checklistener**s**

File: options.go [hunk 8]
```diff
@@ -3,30 +3,20 @@ package gosundheit
...
 	return func(h *health) {
-		if listener != nil {
-			h.checksListener = listener
-		}
+		h.checksListener = listener
 	}
 }
 
-func withDefaultCheckListener() Option {
+func WithHealthListener(listener ...HealthListener) Option {
```
Comment (eranharel) [inline options.go:12]: healthlistener**s**

**Review by eranharel — CHANGES_REQUESTED** [review]

**Comment by eranharel** [comment]
`README.md` needs to be updated too

File: health.go (hunk 3 above)
Comment (bivas) [reply health.go:151]: It created a double lock for no reason.

File: opencencus-check-listener/types.go (hunk 5 above)
Comment (bivas) [reply opencencus-check-listener/types.go:29]: No, as the context will not add that tag

**Review by sagikazarmark — CHANGES_REQUESTED** [review]

File: health_listener.go [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package gosundheit
+
+type HealthListener interface {
```
[the remark this answers is not in the document]
Comment (bivas) [reply health_listener.go:3]: I prefer having each type in it's own file

[the remark this answers is not in the document]
Comment (sagikazarmark) [reply health_listener.go:3]: That's major javaism and not really a go practice, but it's up to you.

File: opencencus-check-listener/check_listener.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"time"
+
+	"go.opencensus.io/stats"
+
+	gosundheit "github.com/AppsFlyer/go-sundheit"
+)
+
+type CheckListener struct {
```
[the remark this answers is not in the document]
Comment (bivas) [reply opencencus-check-listener/check_listener.go]: It is a listener...
I'll rename it to `MetricsListener`

File: utils.go (hunk 2 above)
Comment (bivas) [reply utils.go]: I'm copying the map so the original one won't get changed by listeners

File: README.md [hunk 11]
```diff
@@ -355,12 +380,30 @@ The views can be registered like so:
...
 	"go.opencensus.io/stats/view"
 )
-
-h := gosundheit.New()
+// This listener can act both as check and health listener for reporting metrics
+oc := opencencus.NewMetricsListener()
+h := gosundheit.New(gosundheit.WithCheckListener(oc), gosundheit.WithHealthListener(oc))
 // ...
-view.Register(gosundheit.DefaultHealthViews...)
+view.Register(opencencus.DefaultHealthViews...)
```
Comment (sagikazarmark) [inline README.md:390]: Typo: `s/opencencus/opencensus/`

File: opencencus/go.mod [hunk 12]
```diff
@@ -0,0 +1,14 @@
+module github.com/AppsFlyer/go-sundheit/opencencus
```
Comment (sagikazarmark) [inline opencencus/go.mod:1]: Typo: `opencencus`

File: README.md (hunk 11 above)
Comment (bivas) [reply README.md:390]: 🤦🏼‍♂️ embarrassing ...

File: go.sum [hunk 13]
```diff
@@ -1,4 +1,5 @@
 cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
+github.com/AppsFlyer/go-sundheit v0.2.0/go.mod h1:rCRkVTMQo7/krF7xQ9X0XEF1an68viFR6/Gy02q+4ds=
```
Comment (sagikazarmark) [inline go.sum:2]: This seems a bit strange TBH. Maybe run `go mod tidy`?

File: health.go [hunk 14]
```diff
@@ -46,7 +45,8 @@ func New(opts ...Option) Health {
 type health struct {
 	results        map[string]Result
 	checkTasks     map[string]checkTask
-	checksListener CheckListener
+	checksListener CheckListeners
+	healthListener HealthListeners
```
Comment (sagikazarmark) [inline health.go:49]: I think these should be `CheckListener` and `HealthListener` respectively. `With*Listener` should check if the existing instance in the health struct is already a slice (ie `cl, ok := h.(CheckListeners)`) and append if so, create a slice otherwise.

That way the health struct remains unaware of the implementation.

File: health.go [hunk 15]
```diff
@@ -97,17 +98,24 @@ func (h *health) scheduleCheck(task *checkTask, cfg *Config) {
 		if !h.runCheckOrStop(task, time.After(cfg.InitialDelay)) {
 			return
 		}
-
+		h.reportResults()
```
Comment (sagikazarmark) [inline health.go:101]: It's not immediately obvious why results needs to be reported twice. Maybe add some comments clarifying?

File: health.go [hunk 16]
```diff
@@ -97,17 +98,24 @@ func (h *health) scheduleCheck(task *checkTask, cfg *Config) {
...
 			}
+			h.reportResults()
 		}
 	}()
 }
 
+func (h *health) reportResults() {
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+	h.healthListener.OnResultsUpdated(copyResultsMap(h.results))
```
Comment (sagikazarmark) [inline health.go:116]: In the odd case that this takes slightly longer than expected, I'd copy the results first, release the lock and then call the listener.

File: opencensus/metrics_listener.go [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+	gosundheit "github.com/AppsFlyer/go-sundheit"
+)
+
+// MetricsListener reports metrics on each check registration, start and completion event (as gosundheit.CheckListener)
+// This listener all reports metrics for the entire service health (as gosundheit.HealthListener)
+type MetricsListener struct {
+	classification string
+}
+
+func NewMetricsListener(opts ...Option) *MetricsListener {
```
Comment (sagikazarmark) [inline opencensus/metrics_listener.go:17]: No need for a pointer here IMHO, the struct is not mutated.

File: opencensus/metrics_listener.go [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+func (c *MetricsListener) OnCheckStarted(_ string) {
+}
+
+func (c *MetricsListener) OnCheckCompleted(name string, result gosundheit.Result) {
+	c.recordCheck(name, result)
+}
+
+func (c *MetricsListener) OnResultsUpdated(results map[string]gosundheit.Result) {
+	allHealthy := allHealthy(results)
+	allChecksCtx := createMonitoringCtx(c.classification, ValAllChecks, allHealthy)
```
Comment (sagikazarmark) [inline opencensus/metrics_listener.go:40]: Should probably be a separate PR, but it's worth noting that all checks/all healthy should be an entirely different stat than the individual checks. I'd at least add a todo here.

File: options.go [hunk 19]
```diff
@@ -2,31 +2,22 @@ package gosundheit
 
 type Option func(*health)
 
-// WithCheckListener allows you to listen to check start/end events
-func WithCheckListener(listener CheckListener) Option {
+// WithCheckListeners allows you to listen to check start/end events
+func WithCheckListeners(listener ...CheckListener) Option {
```
Comment (sagikazarmark) [inline options.go:6]: nit: I'd call it `WithCheckListener` regardless the fact that it accepts multiple listeners.

File: options.go [hunk 20]
```diff
@@ -2,31 +2,22 @@ package gosundheit
...
-		if h.checksListener == nil {
-			h.checksListener = noopCheckListener{}
-		}
+		h.healthListener = listener
 	}
 }
 
 // WithDefaults sets all the Health object settings. It's not required to use this as no options is always default
-// Defaults are: no check listener
+// This is a simple placeholder for any future defaults
```
Comment (sagikazarmark) [inline options.go:20]: Suggestion: less code is better code. Drop this until you need it.

**Review by sagikazarmark — CHANGES_REQUESTED** [review]
I had a few more ideas and suggestions, but overall 👍

File: opencensus/metrics_listener.go (hunk 17 above)
Comment (bivas) [reply opencensus/metrics_listener.go:17]: The pointer here is so it will actually return the interface

File: options.go (hunk 19 above)
Comment (bivas) [reply options.go:6]: It was the original method name and then @eranharel asked to use plural

File: options.go (hunk 20 above)
Comment (bivas) [reply options.go:20]: It simplifies the service construction. Less code doesn't apply here.

File: health.go (hunk 14 above)
Comment (bivas) [reply health.go:49]: That will add unnecessary checks to the internal implementation. Simpler, cleaner and easier to follow code documents itself.

File: opencensus/metrics_listener.go (hunk 17 above)
Comment (sagikazarmark) [reply opencensus/metrics_listener.go:17]: I'm not sure I understand. If you remove the pointer receiver from everywhere it'll implement the interface.

File: health.go (hunk 15 above)
Comment (bivas) [reply health.go:101]: This preserves the original behaviour that was here. Each check we run - we report the results - now using the `HealthListener`

File: opencensus/metrics_listener.go (hunk 18 above)
Comment (bivas) [reply opencensus/metrics_listener.go:40]: This preserve the original report that exported by the lib. We should _discuss_ if we wish to change it in some other issue.

File: opencensus/metrics_listener.go (hunk 17 above)
Comment (bivas) [reply opencensus/metrics_listener.go:17]: If you only use the struct, you will need to call the `WithCheckListener` in less fluent way (you're still required to pass the pointer to method receiving the interface) - so easing the usage and returning a pointer.

File: health.go (hunk 14 above)
Comment (sagikazarmark) [reply health.go:49]: I disagree. Currently an implementation is bound to the health struct instead of an interface. IMHO that's worse than a little complexity in a function that's sole purpose is to configure a listener.

File: opencensus/metrics_listener.go (hunk 17 above)
Comment (sagikazarmark) [reply opencensus/metrics_listener.go:17]: > you're still required to pass the pointer to method receiving the interface

My suggestion is to remove the pointer from the method receivers as well. There is no state mutation in this struct, using a pointer is unnecessary and has nothing to do with interface implementation.

File: options.go (hunk 20 above)
Comment (sagikazarmark) [reply options.go:20]: Well, it doesn't do anything, so I'd argue it doesn't simplify anything.

File: health.go (hunk 15 above)
Comment (eranharel) [reply health.go:101]: We need to zero the metrics, and that's why you need an extra event instead of the code we had initially. It won't hurt to have a comment there indeed, cause you asked about this too

Comment (bivas) [reply health.go:101]: It's not for zeroing the metrics, the zero happens in `OnCheckRegister` event. 
This report is because we call `runCheckOrStop` earlier in the flow.

File: opencensus/metrics_listener.go (hunk 18 above)
Comment (eranharel) [reply opencensus/metrics_listener.go:40]: It was a mistake but fixing it will be a breaking change, and I know our dev are using it.
I'd defer this change to later

**Review by eranharel — APPROVED** [review]

**Comment by eranharel** [comment]
Merging, cause the rest of the comments are either minor or controversial.
I like this PR, but I see that we have a lot of work to get this project to the next level.
Thanks @bivas and @sagikazarmark I really appreciate your help

**Review by NeverLiez — COMMENTED** [review]
- [ ]
